planner/core/rule_join_elimination.go

```diff
@@ -0,0 +1,165 @@
+	case LeftOuterJoin:
+		return o.doEliminate(p, 1)
+	case RightOuterJoin:
+		return o.doEliminate(p, 0)
+	default:
+		return nil
+	}
+}
+
+func (o *outerJoinEliminator) doEliminate(p *LogicalJoin, innerChildIdx int) LogicalPlan {
+	// outer join elimination with distinct
+	if len(o.cols) > 0 {
+		cols := o.cols[len(o.cols)-1]
+		allColsInSchema := true
+		for _, col := range cols {
+			columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
+			if c, _ := p.children[1^innerChildIdx].Schema().FindColumn(columnName); c == nil {
+				allColsInSchema = false
+				break
+			}
+		}
+		if allColsInSchema == true {
+			return p.children[1^innerChildIdx]
+		}
+	}
+
+	// outer join elimination without distinct
+	// first, check whether the parent's schema columns are all in left or right
+	if len(o.schemas) == 0 {
+		return nil
+	}
+	for _, col := range o.schemas[len(o.schemas)-1].Columns {
+		columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
+		if c, _ := p.children[1^innerChildIdx].Schema().FindColumn(columnName); c == nil {
+			return nil
+		}
+	}
+	// second, check whether the other side join keys are unique keys
+	p.children[innerChildIdx].buildKeyInfo()
+	var joinKeys []*expression.Column
+	for _, eqCond := range p.EqualConditions {
+		joinKeys = append(joinKeys, eqCond.GetArgs()[innerChildIdx].(*expression.Column))
+	}
+	tmpSchema := expression.NewSchema(joinKeys...)
+	for _, keyInfo := range p.children[innerChildIdx].Schema().Keys {
+		joinKeysContainKeyInfo := true
+		for _, col := range keyInfo {
+			columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
+			if c, _ := tmpSchema.FindColumn(columnName); c == nil {
+				joinKeysContainKeyInfo = false
+				break
+			}
+		}
+		if joinKeysContainKeyInfo {
+			return p.children[1^innerChildIdx]
+		}
+	}
+	// Third, if p.children[innerChildIdx] is datasource, we must check specially index.
+	// Because buildKeyInfo() don't save the index without notnull flag.
+	// But in outer join, null==null don't return true. The null index do not affect the res.
+	if ds, ok := p.children[innerChildIdx].(*DataSource); ok {
+		for _, path := range ds.possibleAccessPaths {
+			if path.isTablePath {
+				continue
+			}
+			idx := path.index
+			if !idx.Unique {
+				continue
+			}
+			joinKeysContainIndex := true
+			for _, idxCol := range idx.Columns {
+				columnName := &ast.ColumnName{Schema: ds.DBName, Table: ds.tableInfo.Name, Name: idxCol.Name}
+				if c, _ := tmpSchema.FindColumn(columnName); c == nil {
+					joinKeysContainIndex = false
+					break
+				}
+			}
+			if joinKeysContainIndex {
+				return p.children[1^innerChildIdx]
+			}
+		}
+	}
+	return nil
+}
+
+func (o *outerJoinEliminator) optimize(p LogicalPlan) (LogicalPlan, error) {
+	// check the distinct
+	if agg, ok := p.(*LogicalAggregation); ok && len(agg.groupByCols) > 0 {
+		isDistinctAgg := true
+		for _, aggDesc := range agg.AggFuncs {
```

Comment:
Can we extract line122 ~ line135 to an individual function like `(o *outerJoinEliminator)isDistinctAgg`

Thus this block can be rewritten as 
``` go
		if o.isDistinctAgg() {
			o.cols = append(o.cols, agg.groupByCols)
			defer func() {
				o.cols = o.cols[:len(o.cols)-1]
			}()
		}
```